homeassistant/components/tesla_wall_connector/__init__.py

```diff
@@ -0,0 +1,185 @@
+"""The Tesla Wall Connector integration."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+import logging
+from typing import Any, Callable
+
+from tesla_wall_connector import WallConnector
+from tesla_wall_connector.exceptions import (
+    WallConnectorConnectionError,
+    WallConnectorConnectionTimeoutError,
+    WallConnectorError,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    CONF_SCAN_INTERVAL_CHARGING,
+    DEFAULT_SCAN_INTERVAL,
+    DEFAULT_SCAN_INTERVAL_CHARGING,
+    DOMAIN,
+    WALLCONNECTOR_CLIENT,
+    WALLCONNECTOR_DATA_LIFETIME,
+    WALLCONNECTOR_DATA_UPDATE_COORDINATOR,
+    WALLCONNECTOR_DATA_VITALS,
+    WALLCONNECTOR_DEVICE_NAME,
+    WALLCONNECTOR_FIRMWARE_VERSION,
+    WALLCONNECTOR_HOST,
+    WALLCONNECTOR_PART_NUMBER,
+    WALLCONNECTOR_SERIAL_NUMBER,
+)
+
+PLATFORMS: list[str] = ["sensor", "binary_sensor"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Tesla Wall Connector from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    hostname = entry.data[CONF_HOST]
+    poll_interval = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+    poll_interval_charging = entry.options.get(
+        CONF_SCAN_INTERVAL_CHARGING, DEFAULT_SCAN_INTERVAL_CHARGING
+    )
+
+    wall_connector = WallConnector(host=hostname, session=async_get_clientsession(hass))
+
+    try:
+        version_data = await wall_connector.async_get_version()
+    except WallConnectorError as ex:
+        raise ConfigEntryNotReady from ex
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        WALLCONNECTOR_CLIENT: wall_connector,
+        WALLCONNECTOR_HOST: hostname,
+        WALLCONNECTOR_PART_NUMBER: version_data.part_number,
+        WALLCONNECTOR_FIRMWARE_VERSION: version_data.firmware_version,
+        WALLCONNECTOR_SERIAL_NUMBER: version_data.serial_number,
+    }
+
```

Comment:
Optional: There is a lot stored in here, consider making this a `dataclass`

Example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/lookin/models.py

Reply:
👍🏻 I was actually planning on making this change :)

Reply:
Made the change in 5a02d54